src/shared/pretty-print.h

```diff
@@ -55,5 +55,23 @@ bool shall_tint_background(void);
 
 void draw_progress_bar(const char *prefix, double percentage);
 void clear_progress_bar(const char *prefix);
-void draw_progress_bar_unbuffered(const char *prefix, double percentage);
-void clear_progress_bar_unbuffered(const char *prefix);
+void draw_progress_bar_impl(const char *prefix, double percentage);
+void clear_progress_bar_impl(const char *prefix);
+
+static inline void fflush_and_disable_bufferingp(FILE **p) {
+        assert(p);
+
+        if (*p) {
+                fflush(*p);
+                setvbuf(*p, NULL, _IONBF, 0); /* Disable buffering again. */
+        }
+}
+
+/* Even though the macro below is slightly generic, but it may not work most streams except for stderr,
+ * as stdout is buffered and fopen() enables buffering by default. */
+#define _WITH_BUFFERED_IO(f, size, p)                                   \
+        _cleanup_(fflush_and_disable_bufferingp) FILE *p = (f);         \
+        (void) setvbuf(p, (char[size]) {}, _IOFBF, size)
```

Comment:
would prefer if this was one expression only, for compat with cases where people omit {} around blocks.

to make that work maybe add a helper that does the _IOFBF thing and then returns the FILE* passed in. That way, we can call that as rhs of the variable declaration.